common/lib/capa/capa/capa_problem.py

```diff
@@ -492,6 +495,121 @@ def get_answer_ids(self):
+            # stop early, and instead of a question we'll report "Question 2".
+            SKIP_ELEMS = ['description']
+            LABEL_ELEMS = ['p', 'label']
+            while questiontext_elem is not None and questiontext_elem.tag in SKIP_ELEMS:
+                questiontext_elem = questiontext_elem.getprevious()
+
+            if questiontext_elem is not None and questiontext_elem.tag in LABEL_ELEMS:
+                question_text = questiontext_elem.text
+            else:
+                # For instance 'd2e35c1d294b4ba0b3b1048615605d2a_2_1' contains 2,
+                # which is used in question number 1
+                question_nr = int(answer_id.split('_')[-2]) - 1
+                question_text = "Question %i" % question_nr
+
+        return question_text
+
+    def find_answer_text(self, answer_id, current_answer_text):
+        """
+        Process a raw answer text to make it more meaningful.
+
+        E.g. in a choice problem like "How much is 2+2?" "Two"/"Three"/"More than three",
+        this function will transform "choice_1" (which is the internal response given by
+        many capa methods) to the human version, e.g. "More than three".
+
+        If the answers are multiple (e.g. because they're from a multiple choice problem),
+        this will join them with a comma.
+
+        If passed a normal string which is already the answer, it doesn't change it.
+
+        Arguments:
+            answer_id: a string like "98e6a8e915904d5389821a94e48babcf_13_1"
+            current_answer_text: a data structure as found in `LoncapaProblem.student_answers`
+                which represents the best response we have until now
+
+        Returns:
+            a string with the human version of the response
+        """
+        if type(current_answer_text) == list:
+            # Multiple answers. This case happens e.g. in multiple choice problems
+            answer_text = ", ".join([
+                self.find_answer_text(answer_id, answer) for answer in current_answer_text
+            ])
+
+        elif current_answer_text.startswith('choice_'):
+            # Many problem (e.g. checkbox) report "choice_0" "choice_1" etc.
+            # Here we transform it
+            elems = self.tree.xpath('//*[@id="{answer_id}"]//*[@name="{choice_number}"]'.format(
+                answer_id=answer_id,
+                choice_number=current_answer_text
+            ))
+            assert len(elems) == 1
+            choicegroup = elems[0].getparent()
+            input_cls = inputtypes.registry.get_class_for_tag(choicegroup.tag)
+            choices_map = dict(input_cls.extract_choices(choicegroup, self.capa_system.i18n, text_only=True))
+            answer_text = choices_map[current_answer_text]
+
+        elif isinstance(current_answer_text, basestring):
+            # Already a string with the answer
+            answer_text = current_answer_text
+
+        else:
+            raise NotImplementedError()
```

Comment:
Hmm... is this final? Will this error be raised for unsupported problem types? Because if so I should catch this error in the report generation code and use the fallback method.

Reply:
This part of the code will (almost) never be reached in the current version of the code. If ``current_answer_text`` is a list, this method will take the first branch. If ``current_answer_text`` is a string, then either this method will take the second or third branch. If it's anything else there will be an error because the object will not have ``startswith``.

Reply:
Sure. Makes sense to raise the error for unsupported types. I will catch the error in my code and use the normal code for generating reports in that case.

Reply:
@clemente I've now adapted my PR to catch this error and fall back to the old way of getting the report.